File: frontend/types/shared-meter.ts
export interface SharedMeter {
  id: number;
  property_id: number;
  name: string;
  room_ids: number[];
}

export interface SharedMeterReadingCreate {
  shared_meter_id: number;
  period: string;
  curr_reading: number;
  prev_reading?: number | null;
}

export interface SharedMeterInvoiceDetail {
  meter_id: number;
  meter_name: string;
  prev_reading: number | null;
  curr_reading: number;
  total_usage: number;
}

export interface SharedMeterReading {
  id: number;
  shared_meter_id: number;
  period: string;
  prev_reading: number | null;
  curr_reading: number;
  is_prev_auto: boolean;
}

File: backend/app/schemas/shared_meter.py
from pydantic import BaseModel, ConfigDict
from decimal import Decimal


class SharedMeterCreate(BaseModel):
    name: str


class SharedMeterUpdate(BaseModel):
    name: str


class SharedMeterRead(BaseModel):
    id: int
    property_id: int
    name: str
    room_ids: list[int] = []


class SharedMeterRoomAdd(BaseModel):
    room_id: int


class SharedMeterReadingCreate(BaseModel):
    shared_meter_id: int
    period: str
    curr_reading: Decimal
    prev_reading: Decimal | None = (
        None  # manual override — used when no previous period exists
    )


class SharedMeterReadingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    shared_meter_id: int
    period: str
    prev_reading: Decimal | None
    curr_reading: Decimal
    is_prev_auto: bool


class SharedMeterInvoiceDetail(BaseModel):
    """Shared meter reading context attached to InvoiceDetailRead."""

    meter_id: int
    meter_name: str
    prev_reading: Decimal | None
    curr_reading: Decimal
    total_usage: Decimal
